### PR TITLE
Changes for SW crypt/digest routines

### PR DESCRIPTION
**Pull request by ifranzki**

### CODE REVIEW

**Comment by ifranzki** [comment]
Fixed.

File: usr/lib/pkcs11/common/sw_crypt.h [hunk 1]
```diff
@@ -23,4 +23,17 @@ CK_RV sw_des3_cbc(CK_BYTE *in_data,
                   CK_ULONG *out_data_len,
                   CK_BYTE *init_v, CK_BYTE *key_value, CK_BYTE encrypt);
 
+#define sw_aes_cbc_encrypt(clear, len, cipher, len2, iv, key, keylen) \
+ sw_aes_cbc(clear, len, cipher, len2, iv, key, keylen, 1)
```
Comment (p-steuer) [inline usr/lib/pkcs11/common/sw_crypt.h]: add parantheses for macro params

File: usr/lib/pkcs11/common/sw_crypt.h [hunk 2]
```diff
@@ -23,4 +23,17 @@ CK_RV sw_des3_cbc(CK_BYTE *in_data,
...
+
+#define sw_aes_cbc_decrypt(clear, len, cipher, len2, iv, key, keylen) \
+ sw_aes_cbc(clear, len, cipher, len2, iv, key, keylen, 0)
```
Comment (p-steuer) [inline usr/lib/pkcs11/common/sw_crypt.h]: add parantheses for macro params

File: usr/lib/pkcs11/common/sw_crypt.c [hunk 3]
```diff
@@ -74,3 +75,35 @@ CK_RV sw_des3_cbc(CK_BYTE *in_data,
...
+    // the aes decrypt will only fail if the data length is not evenly divisible
+    // by AES_BLOCK_SIZE
+    if (in_data_len % AES_BLOCK_SIZE) {
+        TRACE_ERROR("%s\n", ock_err(ERR_DATA_LEN_RANGE));
+        return CKR_DATA_LEN_RANGE;
+    }
+
+    // Encrypt or decrypt the data
+    if (encrypt) {
+        AES_set_encrypt_key(key_value, keylen * 8, &aes_key);
```
Comment (p-steuer) [inline usr/lib/pkcs11/common/sw_crypt.c]: Not using evp interfaces means no guarantee for full platform specific hardware support (old cbc mode apis have full hw support on z though). Dont know if encrypting/decrypting master key is relevant wrt performance.

File: usr/lib/pkcs11/common/utility.c [hunk 4]
```diff
@@ -848,14 +848,18 @@ CK_RV compute_sha(STDLL_TokData_t *tokdata, CK_BYTE *data, CK_ULONG len,
         TRACE_DEBUG("failed to create digest.\n");
         return rv;
     }
-    return sha_hash(tokdata, NULL, FALSE, &ctx, data, len, hash, &hash_len);
+    rv = sha_hash(tokdata, NULL, FALSE, &ctx, data, len, hash, &hash_len);
+
+    if (ctx.context)
+        free(ctx.context);
```
[the remark this answers is not in the document]
Comment (ifranzki) [reply usr/lib/pkcs11/common/utility.c]: Agreed, but usually the context is freed in digest_mgr_cleanup() which is inside dig_mgr.c.  Only in the 2 cases when it is called from utility.c, this is not done that way. I can certainly add a function to mech_sha.c to free the context, and call that one from utility.c, but that would change it only for those 2 cases.

File: usr/lib/pkcs11/common/sw_crypt.c (hunk 3 above)
Comment (ifranzki) [reply usr/lib/pkcs11/common/sw_crypt.c]: I don't think is performance relevant. this code is only used for master key encryption, not for real data encryption. Real data encryption uses the token specific crypto functions.

File: usr/lib/pkcs11/common/mech_sha.c [hunk 5]
```diff
@@ -38,356 +38,23 @@
 #include "tok_spec_struct.h"
 #include "trace.h"
 
-#define SHA_HARDWARE_THRESHHOLD 128000
+#include <openssl/sha.h>
```
Comment (p-steuer) [inline usr/lib/pkcs11/common/mech_sha.c]: Same comment regarding evp as in the aes case applies here. However, for sha1 old interfaces have hw support on z.

File: usr/lib/pkcs11/common/sw_crypt.h (hunk 1 above)
Comment (p-steuer) [reply usr/lib/pkcs11/common/sw_crypt.h]: your right. it doesnt make a difference here.

File: usr/lib/pkcs11/common/mech_sha.c (hunk 5 above)
Comment (ifranzki) [reply usr/lib/pkcs11/common/mech_sha.c]: Also not performance relevant. In general the token provided SHA routines are used. Only if the token does not provide a SHA routine, and its SHA1, then it falls back to the SW implementation. All our tokens do provide SHA routines.

compute_sha1 is used for pin verification, private token object handling and master key handling only. 

compute_sha is used inside mech_rsa for emsa_pss encode and verify, and oaep. Only if SHA1 is used and the token does not provide a SHA implementation, then the SW fallback is used.

File: usr/lib/pkcs11/common/utility.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (ifranzki) [reply usr/lib/pkcs11/common/utility.c]: do you want me to change the 2 cases, or should I leave it as is?

[the remark this answers is not in the document]
Comment (p-steuer) [reply usr/lib/pkcs11/common/utility.c]: what about completing the internal sw_sha1* by an update and final function ? So openssl's sha api is not called one time from sha_hash* layer and the other time from the lower sw_sha1* layer i.e.: sha_hash* > sw_sha1* > SHA1*. Context (de)allocation would be at the sw_sha1 layer:  _sha1 and _sha1_final would free the context and set the pointer to NULL, such digest_mgr_cleanup would not do a double free (cleanup's free would only be needed with unsuccesful sha ops).

File: usr/lib/pkcs11/common/sw_crypt.c (hunk 3 above)
Comment (p-steuer) [reply usr/lib/pkcs11/common/sw_crypt.c]: okay im fine with it.

File: usr/lib/pkcs11/common/mech_sha.c (hunk 5 above)
Comment (p-steuer) [reply usr/lib/pkcs11/common/mech_sha.c]: alright

**Comment by p-steuer** [comment]
Approved , thanks. While your at it could you move the sha_init upwards so it is placed near the other sha* functions and not hidden between the hmac* funcs ?

**Comment by ifranzki** [comment]
moved sha_init upwards

**Pull request merged**